### ISSUE TITLE
HDMI / DVI Specification - VSync/HSync

### ISSUE THREAD

**Issue by yngndrw** [vidor-libraries/VidorFPGA#1]
According to the [HDMI 1.3a Specification - Page 54](https://www.hdmi.org/spec/hdmi1_4b), `VSYNC` and `HSYNC` should only be output for the control period on Channel 0. (I.e. The blue channel) This is also the same according to the [DVI 1.0 Specification - Page 26](http://www.cs.unc.edu/Research/stc/FAQs/Video/dvi_spec-V1_0.pdf).

The [current implementation of the DVI_OUT module](https://github.com/vidor-libraries/VidorFPGA/blob/master/ip/DVI_OUT/DVI_OUT.v#L40-L65) outputs this on all three channels.

Interestingly the HDMI Specification (Page 55) also shows the blanking periods at the start of the frames / lines rather than the end which doesn't seem to match the [FBST module](https://github.com/vidor-libraries/VidorFPGA/blob/master/ip/FBST/FBST.v#L71-L74), but the specification doesn't match the modeline I received from my monitor:
`"1280x720" 74.250 1280 1390 1430 1650 720 725 730 750 +hsync +vsync`

**Comment by pnndra** [comment]
Hi,
Thank you for.your feedback. The IP is really just dvi out, not hdmi as hdmi implementation would require licensing. In dvi the channel 1 and 2 control bits are reserved but can actually be used in user applications, in fact some old dvi transmitter/receiver chip was actually exposing them. In our IP we decided to connect them to syncs but you're right in saying it would be more appropriate to either drive them low or connect them externally for use to drive them.

Regarding fbst the way counters are used is totally arbitrary and it won't affect in any way the output as the timings are equivalent. I can agree on the fact that parameters may be expressed in a more conventional way but I found it useful for readability to start with active data.
In terms of mode line your monitor requests, from memory it seems like SMPTE compliant and you're free to set parameters to whatever you want, however keep In mind that on dvi/hdmi you can produce almost any timing since graphic cards often use various different schemes to calculate timings for video modes and monitors are usually just "suggesting" the preferred modes along with their allowed frequency ranges.
In our case we set timings according to the pixel clock we could achieve with the integer pll we have in our fpga, which is referenced from 24 MHz and can't produce exactly 74.25 MHz so of course to product 1280x720@60Hz we had to change timings. Note also that in the graphics sample fpga we overridden these parameters with a 640x480 mode in order not to consume all the sdram bandwidth.

**Comment by yngndrw** [comment]
Thank you, I didn't realise that the modes were just suggestions. The reason I've been looking at these timings is because I couldn't get the samples to work, with or without modifications - Occasionally I'll see the logo but more often than not the monitor just reports that there's no signal and I've ensured that the FPGA is running the right thing with the correct clock by flashing an LED and timing it.

The monitor also reports a modeline for 720x480 so it should work with the default logo sample, unless it doesn't support 4:3:
```"720x480" 27.000 720 736 798 858 480 489 495 525 -hsync -vsync```

Oddly the sync polarity is different between the two modelines, but I haven't found any documentation on how to interpret these. Any advice on how these polarities related to DVI_OUT would be appreciated.

In my own sample, I tried making a minimal sample using FBST by just passing in 31 bits of '1' as the data - I believe that should work for a white image due to the data enable output handling the blanking as required.

**Comment by pnndra** [comment]
hi, i am assuming that at least you tried the graphics library and it worked for you. if not probably you have another issue with the board.
 
Generally speaking DVI and HDMI monitors are generally not really sensitive to sync polarity and often even to resolutions as long as the image has blanking that allows it to scale it to the monitor timings. depending on the scaler inside your monitor it will eventually reject unknown modes but i'd be surprised if it doesn't accept standard resolutions such as 640x480 or 1280x720.
the biggest issue with monitors is that they usually DON'T like odd vertical frequencies so you should absolutely check that the timing you're using is between 60 and 72 Hz (most monitors won't accept frequencies below 60!).

note also that maximum pixel clock supported by Vidor is 65 MHz as we use LVDS pins which from datasheet have a max rate of 650 mbps, so if you want to output 1280x720 you have to make sure you use a reduced blanking mode that uses 65 MHz pixel clock AND you have to be carefully checking your timings.

finally note that DVI spec requires a minimum pixel clock rate of 25 MHz so any video timing below that fequency needs pixel doubling (or quadrupling in some cases).

regarding clocking i would also suggest that you doublecheck your PLL settings as you need the PLL to generate the pixel clock AND a 5x clock that is used to clock the DVI output core. that clock is needed to shift out data at 10x clock rate by using dual data rate outputs.

**Comment by yngndrw** [comment]
I did try the VidorDrawLogo example from the graphics library first and that didn't work - I assumed that my monitor didn't support the default resolution so that's why I tried my own build with different parameters. It's odd because a couple of times (Twice) I did see the logo, but every other time I just get a lack of signal error so something must be right on the edge.

Thank you for the advice on what these monitors are typically sensitive to and forgiving about - I'll have another play with the parameters and see if I can get a stable image. I'll also double check the PLL settings as you suggest.

**Comment by pnndra** [comment]
Just to make sure... Did you ensure you have the latest bootloader and boot fpga image? If you doubletap the reset button you should revert to boot fpga and that should show the Arduino logo.
In addition to this, if you load a sample sketch please make sure you open the serial terminal as the samples wait for serial to be connected prior to initializing fpga. Before that everything is just dark...

**Comment by yngndrw** [comment]
I have Arduino SAMD Boards v1.8.4 installed and the SAMD_BootloaderUpdater tool reported that the bootloader is up to date.

The VidorGraphics and VidorPeripherals libraries are both v1.1.0.

The VidorTestSketch reported: `Vidor bitstream version: 1020109`

VidorDrawLogo doesn't output anything to the screen even after opening the Serial Monitor. I confirmed that `FPGA.begin()` is succeeding by writing to the serial port afterwards.

Double tapping the reset button also didn't seem to show the logo, even with the Serial Monitor open.


I'll confirm the timings tomorrow, but sadly it looks like the standard examples are not working in my case.

**Comment by pnndra** [comment]
Hmm this is weird. I would suggest you try this on another monitor and if it still doesn't work you file for a replacement. Maybe you want to try first with the boot fpga recovery procedure as in the forum sticky post...

**Comment by yngndrw** [comment]
Sadly the boot FPGA recovery procedure didn't make a difference on this monitor.

I've just tried it three different screens:
1) A 1080p LG TV (Pre WebOS) - Works perfectly, very responsive when turning the Vidor4000 on and off. (Although 480p doesn't look great at this screen size!)
2) My PC monitor (21:9 ultrawide @ 3440x1440 with G-Sync, so I didn't expect this to work too well) - This does display an image, but it flickers a lot.
3) The little 7" screen (Fieldview 777SB, 800x480 native: https://hobbyking.com/en_us/7-sun-readable-1000cd-m2-dvr-fpv-monitor-uk-plug.html) - Just a black screen except occasionally (Twice) when it starts working and gives a good solid image in those circumstances.

The ultrawide monitor only gives modelines at its native resolution (Including this! `"3440x1440" 515.000 3440 3488 3520 3600 1440 1443 1453 1506 +hsync -vsync`), but it does also output the following so it should be happy:
```
Standard timings supported
     640 x  480p at  60Hz - IBM VGA
     800 x  600p at  60Hz - VESA
    1024 x  768p at  60Hz - VESA
```

I'm not too worried about the ultrawide as it really isn't designed for this sort of thing, but I'd really like to get it working on the Fieldview 777SB. I still need to double-check and play with the timings so I'll do that next.

**Comment by yngndrw** [comment]
Turns out that it was the pixel rate. (Or the refresh rate, I don't know - they are all related) I noticed that the EDID data for the Fieldview 777SB lists the "Vertical scan range" as "50-75Hz" yet the ultrawide (Which nearly displays a stable image) lists a range of "30-95Hz" - Leading me to belive that the Fieldview 777SB is less tolerant.

The default pixel rate of 24MHz gives a line rate of just under 28KHz and refresh rate of 53.28Hz, both well within the stated allowed timings for 720x480p. (Which I've switched to in my own example)

The desired pixel rate of 27MHz isn't achievable with the 48MHz input clock, so instead I tried 30MHz / 150MHz on the PLL. This gives a line rate of just under 35KHz and a refresh rate of 66.6Hz - And displays my white screen as desired.

Switching to the standard 640x480 timings provided in the Vidor 4000 (But with the 30MHz pixel rate) gives my white screen with black vertical bars as you'd expect.

Sadly it still doesn't work on the ultrawide, but I'm going to put that down to G-Sync - I don't care about it working on that one anyway.

It's a little strange that it didn't work at 24MHz, I even checked the clock output of the SAMD21 but it was 48MHz according to my oscilloscope - That is not a fun pin to probe!

Anyway thank you for your time and help.

Full EDID for reference:
```
Monitor
  Model name............... 32V3H-H6A
  Windows description...... Generic PnP Monitor
  Manufacturer............. RTK
  Plug and Play ID......... RTK4C54
  Serial number............ 1
  Manufacture date......... 2007, ISO week 12
  Filter driver............ None
  -------------------------
  EDID revision............ 1.3
  Input signal type........ Digital (DVI)
  Color bit depth.......... Undefined
  Display type............. RGB color
  Screen size.............. 700 x 390 mm (31.5 in)
  Power management......... Standby
  Extension blocs.......... 1 (Unknown type)
  -------------------------
  DDC/CI................... n/a

Color characteristics
  Default color space...... Non-sRGB
  Display gamma............ 2.20
  Red chromaticity......... Rx 0.650 - Ry 0.330
  Green chromaticity....... Gx 0.290 - Gy 0.610
  Blue chromaticity........ Bx 0.150 - By 0.070
  White point (default).... Wx 0.272 - Wy 0.275
  Additional descriptors... None

Timing characteristics
  Horizontal scan range.... 24-60kHz
  Vertical scan range...... 50-75Hz
  Video bandwidth.......... 110MHz
  CVT standard............. Not supported
  GTF standard............. Not supported
  Additional descriptors... None
  Preferred timing......... Yes
  Native/preferred timing.. 1280x720p at 60Hz (16:9)
    Modeline............... "1280x720" 74.250 1280 1390 1430 1650 720 725 730 750 +hsync +vsync
  Detailed timing #1....... 720x480p at 60Hz (16:9)
    Modeline............... "720x480" 27.000 720 736 798 858 480 489 495 525 -hsync -vsync

Standard timings supported
     720 x  400p at  70Hz - IBM VGA
     640 x  480p at  60Hz - IBM VGA
     640 x  480p at  72Hz - VESA
     640 x  480p at  75Hz - VESA
     800 x  600p at  56Hz - VESA
     800 x  600p at  60Hz - VESA
     800 x  600p at  72Hz - VESA
     800 x  600p at  75Hz - VESA
    1024 x  768p at  60Hz - VESA
    1024 x  768p at  70Hz - VESA
    1024 x  768p at  75Hz - VESA
    1280 x 1024p at  75Hz - VESA
    1440 x  900p at  60Hz - VESA STD
    1440 x  900p at  75Hz - VESA STD
    1440 x  900p at  85Hz - VESA STD

Report information
  Date generated........... 07/12/2019
  Software revision........ 2.91.0.1043
  Data source.............. Registry - NB: improperly installed
  Operating system......... 10.0.18362.2

Raw data
  00,FF,FF,FF,FF,FF,FF,00,4A,8B,54,4C,01,00,00,00,0C,11,01,03,81,46,27,78,8A,A5,8E,A6,54,4A,9C,26,
  12,45,46,AF,CF,00,95,00,95,0F,95,19,01,01,01,01,01,01,01,01,01,01,01,1D,00,72,51,D0,1E,20,6E,28,
  55,00,B9,88,21,00,00,1E,8C,0A,D0,8A,20,E0,2D,10,10,3E,96,00,B9,88,21,00,00,18,00,00,00,FD,00,32,
  4B,18,3C,0B,00,0A,20,20,20,20,20,20,00,00,00,FC,00,33,32,56,33,48,2D,48,36,41,0A,20,20,20,01,29
```

**Comment by pnndra** [comment]
Hi Andrew,
i guess you just raised the root cause of the issue. i wrote it above but didn't recognize it earlier. DVI spec requires pixel clock of at least 25 MHz and we're clocking at 24. for sure we need to update this to a different value as it is (slightly) out of spec and some displays as yours may not work.
we were trying to keep the pixel clock as low as possible as that also impacts on memory bandwidth used in the SDRAM and that is really precious. 
for the PLL i would recommend to change it to 25 MHz rather than 30. that should work and quartus will be able to implement it if you set it as follows:
C0: 25 MHz (pixel clock)
C1: 125 MHz (5x pixel clock)
C2: 100 MHz (MEM Clk)
C3: 100 MHz (SDRAM Clk)
C4: 50 MHz (Flash clock)

you probably were not able to have it working because if you leave Flash Clock at 60 MHz there's no common denominator that can be used to achieve the output frequencies.
with 25 MHz and original fbst timings you should get 62 Hz. you can get to 60 Hz by just setting HTOTAL to 794 instead of 762.

**Comment by yngndrw** [comment]
Sorry I didn't put two and two together when you mentioned it earlier, I think I just took the flexibility of some displays for granted when I saw it working well on the TV and then forgot about it.

I can confirm that the PLL / HTOTAL changes that you suggested are working great on this display, at least in my example with the `DVI_OUT` & `FBST` modules. The ID `RTK4C54` seems to match a few monitors so I suspect this would improve compatibility with a number of displays, I did notice a few older posts on the forums about compatibility with various monitors so hopefully this will resolve those issues as well.

I haven't been able to add the SDRAM to my example due to this issue (https://github.com/vidor-libraries/VidorBitstream/issues/5 - Sorry if it's in the wrong repository, I wasn't sure which was best to raise it within), but I think it's fair to assume it would work fine with that.
